Refactor constraints and add feature groups

s9latimm opened the pull request.

s9latimm committed:
WIP: Add node class

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
A	include/vara/Feature/Group.h
A	include/vara/Feature/Node.h
M	include/vara/Feature/OrderedFeatureVector.h
M	lib/Feature/Feature.cpp
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
WIP: Fix bindings

M	bindings/python/vara-feature/pybind_Feature.cpp

Kallistos committed:
constraint tree structure added, builder is missing

Kallistos committed:
Merge branch 'f-FMConstraints' of https://github.com/se-passau/vara-feature into f-FMConstraints

s9latimm committed:
WIP: Refactor feature tree

M	bindings/python/vara-feature/pybind_Feature.cpp
M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
A	include/vara/Feature/FeatureTreeNode.h
D	include/vara/Feature/Node.h
M	include/vara/Feature/OrderedFeatureVector.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	tools/fm-viewer/FeatureModelViewer.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelWriter.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
Clean up, added TODO's

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureTreeNode.h
M	include/vara/Feature/Relationship.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelWriter.cpp

s9latimm committed:
Make inheritence private

M	include/vara/Feature/FeatureTreeNode.h

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -0,0 +1,64 @@
+
+private:
+  friend class FeatureModel;
+  friend class FeatureModelBuilder;
```
Why do constraints need private member access to a FeatureModel or the Builder?

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -0,0 +1,64 @@
+        Constraint(ConstraintKind::CK_UNARY) {}
+
+private:
+  string Operator;
```
Why are Operators strings?

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -0,0 +1,64 @@
+        RightOperand(RightOperand), Constraint(ConstraintKind::CK_BINARY) {}
+
+private:
+  string Operator;
```
See, below

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -0,0 +1,64 @@
+
+/// \brief Tree like representation of constraints between features
+
+class Constraint {
```
I dont fully understand this constraint design. There is no public interface that users can use so what is the planed purpose of the Constraint class?

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -51,109 +48,15 @@ class Feature {
-  [[nodiscard]] const_feature_iterator begin() const {
-    return children_begin();
+  [[nodiscard]] Feature *getParentFeature() const {
+    for (auto *P = this->getParent(); P; P = P->getParent()) {
```
```suggestion
    for (FeatureTreeNode *P = this->getParent(); P; P = P->getParent()) {
```
It's not quite clear here with auto. One could expect here that getParent returns a `Feature`.

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -51,109 +48,15 @@ class Feature {
-    return children_begin();
+  [[nodiscard]] Feature *getParentFeature() const {
+    for (auto *P = this->getParent(); P; P = P->getParent()) {
+      auto *F = llvm::dyn_cast<Feature>(P);
+      if (F) {
+        return F;
+      }
```
```suggestion
      if (llvm::isa<Feature>(P)) {
        return P;
      }
```

s9latimm replied:
This method should return an actual parent `Feature` (if exists), but `P` is only a `FeatureTreeNode` at this point.

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -51,109 +48,15 @@ class Feature {
+        return F;
+      }
+    }
+    return nullptr;
```
```suggestion
    assert(false && "Node had no parent but was not the root node.");
    return nullptr;
```
related to above

s9latimm replied:
If a node has no parent `isRoot` always returns true. See my comment regarding checking this inside the loop. Currently we guarantee consistency only when inside a FMTree (with a single root).

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -51,109 +48,15 @@ class Feature {
+      if (F) {
+        return F;
+      }
+    }
```
```suggestion
      if (P->isRoot()) {
        return nullptr;
      }
    }
```
We could add an additional safty check here, not sure if it is worth it. What do you think?

s9latimm replied:
Current implementation of `isRoot` just checks `Parent == nullptr`, so one would already expect, that `getParent` on a root node returns a  `nullptr` (which is handled in the loop condition).

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -21,7 +23,7 @@ class FeatureModel {
   using OrderedFeatureTy = OrderedFeatureVector;
-  using ConstraintTy = llvm::SmallVector<std::pair<std::string, bool>, 3>;
+  using ConstraintTy = Constraint;
   using ConstraintsTy = std::vector<ConstraintTy>;
```
```suggestion
  using ConstraintContainerTy = std::vector<ConstraintTy>;
```
I think this could be a better name for what we do with the type.

vulder commented on include/vara/Feature/FeatureTreeNode.h:
```diff
@@ -0,0 +1,88 @@
+public:
+  enum class NodeKind { NK_FEATURE, NK_GROUP };
+
+  using NodeSetType = typename std::set<FeatureTreeNode *>;
```
Hmm, if we expect only a few child nodes we could use a SmallSet here with an appropriate size.

vulder commented on include/vara/Feature/FeatureTreeNode.h:
```diff
@@ -0,0 +1,88 @@
+  [[nodiscard]] iterator end() { return DGNode::end(); }
+  [[nodiscard]] const_iterator end() const { return DGNode::end(); }
+
+  llvm::iterator_range<iterator> children() {
```
```suggestion
  [[nodiscard]] llvm::iterator_range<iterator> children() {
```

vulder commented on include/vara/Feature/OrderedFeatureVector.h:
```diff
@@ -32,8 +32,11 @@ class OrderedFeatureVector {
   template <typename T> void insert(llvm::iterator_range<T> FF) {
-    for (const auto &F : FF) {
-      insert(F);
+    for (const auto &N : FF) {
```
`FF` and you get a `N` when iterating over it. We could use better names here :smile:

vulder commented on include/vara/Feature/Relationship.h:
```diff
@@ -0,0 +1,46 @@
+  Relationship(RelationshipKind Kind)
+      : Kind(Kind), FeatureTreeNode(NodeKind::NK_GROUP) {}
+
+  [[nodiscard]] RelationshipKind getRelationshipKind() const { return Kind; }
```
I, normaly, use just `getKind` for kind accessers. I think this should be the llvm way, but we should double check to be sure.

vulder commented on include/vara/Feature/FeatureTreeNode.h:
```diff
@@ -0,0 +1,88 @@
+  FeatureTreeNode &operator=(const FeatureTreeNode &) = delete;
+  virtual ~FeatureTreeNode() = default;
+
+  [[nodiscard]] NodeKind getNodeKind() const { return Kind; }
```
Same as below `getKind`

vulder commented on lib/Feature/Feature.cpp:
```diff
@@ -11,24 +11,15 @@ std::string Feature::toString() const {
-  for (const auto &Alternative : Alternatives) {
-    StrS << Alternative->Name << ",";
+  for (const auto &Child : children()) {
+    auto *F = llvm::dyn_cast<Feature>(Child);
+    if (F) {
```
use `llvm::isa`

s9latimm replied:
I call `F->getName()` afterwards, which needs this `Feature`.

vulder replied:
A yes, I missed that.

vulder commented on lib/Feature/FeatureModel.cpp:
```diff
@@ -18,73 +18,74 @@ bool FeatureModel::addFeature(std::unique_ptr<Feature> Feature) {
       Root = Value;
     } else {
-      Value->setParent(Root);
+      Value->setParent(*Root);
```
If we have a lot of these uses we could add an setParent overload for pointers.

vulder commented on lib/Feature/FeatureModel.cpp:
```diff
@@ -167,35 +169,44 @@ bool FeatureModelBuilder::addFeature(Feature &F) {
-            std::string(F.getName()), dynamic_cast<NumericFeature *>(&F)->getValues(),
-            F.isOptional(), Loc)) {
+            std::string(F.getName()),
+            dynamic_cast<NumericFeature *>(&F)->getValues(), F.isOptional(),
```
Hmm, I'm wondering aren't we using the llvm kind/classof infra for Feature? So, this shoudl/could be a `llvm::dyn_cast`?

s9latimm replied:
It should have been. :laughing:

vulder commented on lib/Feature/FeatureModel.cpp:
```diff
@@ -107,8 +108,8 @@ bool FeatureModelBuilder::buildTree(const string &FeatureName,
     }
-    Features[FeatureName]->addChild(Features[Child].get());
-    Features[Child]->setParent(Features[FeatureName].get());
+    Features[FeatureName]->addEdge(*Features[Child]);
```
Same here

vulder reviewed (changes requested):
I did a first pass over the current changes, looks good overall, I like the llvm DGNode choice.

s9latimm committed:
Suggested changes

M	.travis.yml
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureTreeNode.h
M	include/vara/Feature/OrderedFeatureVector.h
M	include/vara/Feature/Relationship.h
M	lib/Feature/FeatureModel.cpp

Kallistos committed:
Added ConstraintOperator enum, to replace string representation (eases parsing), deleted friends in Constraints

Kallistos committed:
constraint tree, added classes for operands to extend type information of the tree nodes

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -25,55 +25,166 @@ class Constraint {
-    default: return "";
-    }
-  };
+class BoolschesConstraint {
```
```suggestion
class BooleanConstraint {
```
This should be spell in English I guess :)

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -25,55 +25,166 @@ class Constraint {
+      : LeftOperand(LeftOperand),
         RightOperand(RightOperand), Constraint(ConstraintKind::CK_BINARY) {}
 
+  [[nodiscard]] virtual std::string toString() const;
```
```suggestion
  [[nodiscard]] std::string toString() const override;
```

Kallistos replied:
If i change this to 
```suggestion
  [[nodiscard]] virtual std::string toString() const;
```
there is no virtual function, which means the class is not abstract, right?

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -25,55 +25,166 @@ class Constraint {
+  UnaryConstraint(Constraint *Operand) : Operand(Operand),
         Constraint(ConstraintKind::CK_UNARY) {}
 
+  [[nodiscard]] virtual std::string toString() const;
```
```suggestion
  [[nodiscard]] std::string toString() const override;
```

Kallistos replied:
same as above

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -25,55 +25,166 @@ class Constraint {
   ConstraintKind Kind;
 };
```
Github does not let me annotate the lines above :cry: 
Both sub hierachies have the `toString` method so should be not put it into the base class?

Add this to the public API
```cpp
[[nodiscard]] virtual std::string toString() const;
```

Kallistos replied:
you're right! but how can i mark both subclasses as abstract? (BooleanConstraint and NumericConstraint?)

vulder replied:
As long as they don't offer a `toString` method them selfs (or higher up the hierarchy) they are still abstract.

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -25,55 +25,166 @@ class Constraint {
+  [[nodiscard]] std::string toString() const override;
+};
+
+class XOrConstraint : public BinaryConstraint, public BoolschesConstraint {
```
```suggestion
class XorConstraint : public BinaryConstraint, public BoolschesConstraint {
```
Maybe spellit `Xor` or `XOR`, spelling it `XOr` looks very strange to me :laughing:

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -25,55 +25,166 @@ class Constraint {
+  [[nodiscard]] std::string toString() const override;
+};
+
 class PrimaryConstraint : public Constraint {
```
What is the reasoning behind `Primary` as a constraint name? I'm not familiar with this.

Kallistos replied:
Primary represents the leave nodes in the constraint tree, i based this on the expression hierarchy from C; there is a primary expression either a constant, a literal and so on. In our case this is right now either a number or  a feature.

Kallistos committed:
annotated changes from @vulder in the pull request

s9latimm committed:
Update FMBuilder with basic constraints

M	.travis.yml
M	CMakeLists.txt
M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/CMakeLists.txt
A	lib/Feature/Constraint.cpp
M	lib/Feature/FeatureModel.cpp

Kallistos committed:
merge conflict solved

vulder commented on .travis.yml:
```diff
@@ -12,43 +12,43 @@ jobs:
+      stage: Legacy
       addons: *clang10
-      env: CXX=clang++-10 BUILD_TYPE=Release
+      env: CXX=clang++-9 BUILD_TYPE=Release LLVM_VERSION=10
```
```suggestion
      env: CXX=clang++-10 BUILD_TYPE=Release LLVM_VERSION=10
```

Kallistos committed:
renamed XOr (->Xor)

s9latimm committed:
Add override

M	include/vara/Feature/Constraint.h

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -96,8 +97,11 @@ class FeatureModel {
+  ConstraintContainerTy Constraints;
   Feature *Root;
 
+  FeatureModel() = default;
```
Why do we add a default ctor for FeatureModels? Before we always wanted to have at least a path and a name.

s9latimm replied:
This way `FeatureModelBuilder` can inherit the data structure without already setting publicly required fields (like path / name) as they are build afterwards.

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -37,161 +45,216 @@ class NumericConstraint {
-  Feature *Feature;
+  friend FeatureModelBuilder;
+
+  std::variant<Feature *, std::string> FV;
```
What is the reasoning behind the string alternative?

s9latimm replied:
While parsing  / building the constraint tree, the actual feature might not be present in the FM, so the name is used as placeholder and replaced afterwards. On the other hand we are able to build a constraint tree without requiring any feature parsing.

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -37,161 +45,216 @@ class NumericConstraint {
+  PrimaryConstraint(std::variant<Feature *, std::string> FV)
+      : FV(std::move(FV)), Constraint(ConstraintKind::CK_PRIMARY) {}
+
+  std::variant<Feature *, std::string> getFeature() { return FV; }
```
This interface seems a bit off. One requests a Feature but also could get a string (currently I'm not sure what the meaning of the string is). Depending on the string meaning, it could be better to check the type in the variant here and return an optional feature (or nullptr) if there was a string in the variant and the `Feature *` otherwise. We could then offer the same interface for the "meaning of string".

This would not tie the user into our variant usage and require all using code to make variant dispatches.

s9latimm replied:
Indeed, it seems reasonable to wrap the optional feature within the getter.

Kallistos committed:
removed unnecassary toString in abstractClasses Binary and Unary Constraint

s9latimm committed:
Add backref

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h

s9latimm committed:
Merge branch 'f-FMConstraints' of github.com:se-passau/vara-feature into f-FMConstraints

M	include/vara/Feature/Constraint.h

s9latimm committed:
Remove toString

M	include/vara/Feature/Constraint.h

s9latimm committed:
Fix travis

M	.travis.yml

s9latimm committed:
Update variant

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/Constraint.cpp

s9latimm committed:
Update primay

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/Constraint.cpp

s9latimm committed:
Make abstract

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h

Kallistos committed:
implemented toString for constraint classes

Kallistos committed:
Merge branch 'f-FMConstraints' of https://github.com/se-passau/vara-feature into f-FMConstraints

s9latimm committed:
Print constraints

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/Constraint.cpp
M	lib/Feature/FeatureModelParser.cpp

s9latimm committed:
Add relationship container

M	include/vara/Feature/FeatureModel.h
M	lib/Feature/FeatureModel.cpp

s9latimm committed:
Add tests

M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureTreeNode.h
M	lib/Feature/FeatureModel.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/CMakeLists.txt
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/NumericFeature.cpp
A	unittests/Feature/Relationship.cpp

vulder commented on .travis.yml:
```diff
@@ -13,42 +13,42 @@ jobs:
+            - llvm-11-dev
+            - clang++-11
+            - clang-tidy-11
+      env: CXX=clang++-11 BUILD_TYPE=Debug LLVM_VERSION=11
```
Is the `LLVM_VERSION=11` needed? Normally, it should default to the current one, meaning be automatically 11.

vulder commented on bindings/python/vara-feature/pybind_Feature.cpp:
```diff
@@ -24,62 +24,67 @@ void init_feature_module_feature(py::module &M) {
+      //                    py::return_value_policy::reference,
+      //                    R"pbdoc(Returns the code location of the feature
+      //                    variable if
+      //                possible, otherwise, `None`.)pbdoc")
```
Do we have new iterators for these? Or should this be removed?

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -0,0 +1,379 @@
+
+class PrimaryIntegerConstraint : public PrimaryConstraint {
+public:
+  PrimaryIntegerConstraint(int V) : V(V) {}
```
needs member accessor

vulder replied:
probably needs kind

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -0,0 +1,379 @@
+/// \brief Tree like representation of constraints between features
+class Constraint {
+public:
+  enum class ConstraintKind { CK_BINARY, CK_UNARY, CK_PRIMARY };
```
create specific kinds for sub types

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -33,7 +33,28 @@ bool FeatureModelXmlParser::parseConfigurationOption(xmlNode *Node,
               std::unique_ptr<xmlChar, void (*)(void *)> CCnt(
                   xmlNodeGetContent(Child), xmlFree);
-              FMB.addExclude(Name, reinterpret_cast<char *>(CCnt.get()));
+              FMB.addConstraint(std::make_unique<ImpliesConstraint>(
```
Use extra `ExcludeConstraint`

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -106,17 +129,18 @@ bool FeatureModelXmlParser::parseConstraints(xmlNode *Node) {
                                          .get()));
         const std::regex Regex(R"((!?\w+))");
```
Use llvm::Regex

vulder commented on include/vara/Feature/FeatureTreeNode.h:
```diff
@@ -0,0 +1,91 @@
+
+protected:
+  FeatureTreeNode(NodeKind Kind, FeatureTreeNode *Parent,
+                  const NodeSetType &Children)
```
```suggestion
                  const llvm::SmallSetImpl<FeatureTreeNode *> &Children)
```

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -0,0 +1,379 @@
+  virtual void accept(ConstraintVisitor &V) = 0;
+
+private:
+  ConstraintKind Kind;
```
```suggestion
  const ConstraintKind Kind;
```
So we can guarantee that the kind does not change during the Lifetime of a constraint.

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -0,0 +1,379 @@
+
+  Constraint(ConstraintKind Kind) : Kind(Kind) {}
+  Constraint(const Constraint &) = delete;
+  Constraint &operator=(const Constraint &) = delete;
```
```suggestion
  Constraint &operator=(const Constraint &) = delete;
  Constraint(Constraint &&) = delete;
  Constraint &operator=(Constraint &&) = delete;
```
If we don't implement massign and mctor our selfs they will be disabled, so we should at least flag them as deleted here, or implement them.

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -188,39 +107,28 @@ class Feature {
-
-  void addImplication(Feature *F) { Implications.insert(F); }
-
   FeatureKind Kind;
```
```suggestion
  const FeatureKind Kind;
```

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -188,39 +107,28 @@ class Feature {
-
   FeatureKind Kind;
   string Name;
   bool Opt;
```
I would move the `bool` flag to the end of the member list.

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -21,25 +22,27 @@ using std::string;
+      : FeatureTreeNode(NodeKind::NK_FEATURE), Kind(FeatureKind::FK_UNKNOWN),
+        Name(std::move(Name)), Opt(false), Source(std::nullopt) {}
   Feature(const Feature &) = delete;
   Feature &operator=(const Feature &) = delete;
```
Same with the mctor/massign here.

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -180,6 +91,14 @@ class Feature {
     Source = std::move(FeatureSR);
   }
 
+  llvm::iterator_range<constraint_iterator> constraints() {
```
```suggestion
  [[nodiscard]] llvm::iterator_range<constraint_iterator> constraints() {
```

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -357,9 +323,14 @@ template <> struct GraphWriter<vara::feature::FeatureModel *> {
-                 llvm::formatv("arrowhead={0}",
-                               Child->isOptional() ? "odot" : "dot"));
+        auto *F = llvm::dyn_cast<vara::feature::Feature>(Child);
+        if (F) {
```
```suggestion
        
        if (auto *F = llvm::dyn_cast<vara::feature::Feature>(Child); F) {
```
If you only use the value inside the if, you can use a if initializer. Makes for better scoping. `F` will be out of scope after the `if`, making it impossible to be used after wards, which would be good in the case `F == nullptr`

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -371,19 +342,47 @@ template <> struct GraphWriter<vara::feature::FeatureModel *> {
-
+    std::string Label;
+    auto *F = llvm::dyn_cast<vara::feature::Feature>(Node);
+    if (F) {
```
```suggestion
    if (auto *F = llvm::dyn_cast<vara::feature::Feature>(Node); F) {
```

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -371,19 +342,47 @@ template <> struct GraphWriter<vara::feature::FeatureModel *> {
+              "cellpadding=\"5\">{0}{1}{2}</table>>",
+              llvm::formatv("<tr><td><b>{0}</b></td></tr>",
+                            DOT::EscapeString(F->getName().str()))
+                  .str(),
```
```suggestion
                  ,
```

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -371,19 +342,47 @@ template <> struct GraphWriter<vara::feature::FeatureModel *> {
+                                       F->getFeatureSourceRange()->toString()))
+                         .str()
+                   : ""))
+              .str();
```
```suggestion
              ;
```
I think we don't need to convert these objects to strings before using them. I'm not 100% sure so please double check :)

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -371,19 +342,47 @@ template <> struct GraphWriter<vara::feature::FeatureModel *> {
+              .str();
+    } else {
+      auto *R = llvm::dyn_cast<vara::feature::Relationship>(Node);
+      if (R) {
```
```suggestion
      if (auto *R = llvm::dyn_cast<vara::feature::Relationship>(Node); R) {
```

vulder commented on lib/Feature/Feature.cpp:
```diff
@@ -11,24 +11,15 @@ std::string Feature::toString() const {
-    StrS << Alternative->Name << ",";
+  for (const auto &Child : children()) {
+    auto *F = llvm::dyn_cast<Feature>(Child);
+    if (F) {
```
Some here

vulder commented on lib/Feature/FeatureModel.cpp:
```diff
@@ -166,35 +154,36 @@ bool FeatureModelBuilder::addFeature(Feature &F) {
-  for (const auto *Implication : F.implications()) {
-    this->addConstraint({{std::string(F.getName()), false}, {std::string(Implication->getName()), true}});
+    const auto *C = llvm::dyn_cast<vara::feature::Feature>(Child);
+    if (C) {
```
same here

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -33,7 +33,28 @@ bool FeatureModelXmlParser::parseConfigurationOption(xmlNode *Node,
+              FMB.addConstraint(std::make_unique<ImpliesConstraint>(
+                  std::make_unique<PrimaryFeatureConstraint>(
+                      std::make_unique<Feature>(Name)),
+                  std::make_unique<NotConstraint>(
```
```suggestion
                  make_unique<NotConstraint>(
```
Add a using decl at the beginning of the file.
```
using std::make_unique;
```
This should make the code shorter and a bit easier to read.

vulder commented on lib/Feature/FeatureModelParser.cpp:
```diff
@@ -96,7 +117,9 @@ bool FeatureModelXmlParser::parseOptions(xmlNode *Node, bool Num = false) {
 }
 
-bool FeatureModelXmlParser::parseConstraints(xmlNode *Node) {
+// TODO(s9latimm): Refactor with new Constraints representation
```
If you want to keep the TODO for future reference, please relate it to the issue that implements the constraint parser.

```
// TODO(se-passau/VaRA#666): Comment
```

vulder commented on lib/Feature/FeatureModelWriter.cpp:
```diff
@@ -140,7 +140,10 @@ int FeatureModelXmlWriter::writeNumericFeatures(xmlTextWriterPtr Writer) {
 }
 
-int FeatureModelXmlWriter::writeBooleanConstraints(xmlTextWriterPtr Writer) {
+// TODO(s9latimm): Refactor with new Constraints representation
```
same

vulder commented on lib/Feature/FeatureModelWriter.cpp:
```diff
@@ -223,39 +225,41 @@ int FeatureModelXmlWriter::writeFeature(xmlTextWriterPtr Writer,
     CHECK_RC
   }
 
+  // TODO(s9latimm): Refactor with new Constraints representation
```
same

vulder commented on include/vara/Feature/FeatureTreeNode.h:
```diff
@@ -0,0 +1,91 @@
+  bool addEdge(FeatureTreeNode *Feature) { return DGNode::addEdge(*Feature); }
+
+  FeatureTreeNode *Parent{nullptr};
+  NodeKind Kind;
```
```suggestion
  const NodeKind Kind;
```

vulder commented on include/vara/Feature/FeatureTreeNode.h:
```diff
@@ -0,0 +1,91 @@
+
+  [[nodiscard]] bool isRoot() const { return Parent == nullptr; }
+
+  bool hasEdgeFrom(FeatureTreeNode &N) const { return Parent == &N; }
```
```suggestion
  [[nodiscard]] bool hasEdgeFrom(FeatureTreeNode &N) const { return Parent == &N; }
```

vulder commented on include/vara/Feature/FeatureTreeNode.h:
```diff
@@ -0,0 +1,91 @@
+
+  FeatureTreeNode(NodeKind Kind) : Kind(Kind){};
+  FeatureTreeNode(const FeatureTreeNode &) = delete;
+  FeatureTreeNode &operator=(const FeatureTreeNode &) = delete;
```
mctor/massign

vulder commented on include/vara/Feature/FeatureTreeNode.h:
```diff
@@ -0,0 +1,91 @@
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <llvm/ADT/SetVector.h>
```
```suggestion
#include "llvm/ADT/SetVector.h"

```
+ include sorting

vulder commented on include/vara/Feature/OrderedFeatureVector.h:
```diff
@@ -31,9 +31,12 @@ class OrderedFeatureVector {
-      insert(F);
+  template <typename T> void insert(llvm::iterator_range<T> Iter) {
+    for (const auto &Ptr : Iter) {
+      auto *F = llvm::dyn_cast<Feature>(Ptr);
+      if (F) {
```
same

vulder commented on include/vara/Feature/Relationship.h:
```diff
@@ -0,0 +1,46 @@
+  }
+
+private:
+  RelationshipKind Kind;
```
```suggestion
  const RelationshipKind Kind;
```

vulder commented on include/vara/Feature/Relationship.h:
```diff
@@ -0,0 +1,46 @@
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <llvm/ADT/SetVector.h>
```
```suggestion
#include "llvm/ADT/SetVector.h"

```
+ include sorting

vulder commented on include/vara/Feature/Relationship.h:
```diff
@@ -0,0 +1,46 @@
+#include <utility>
+#include <variant>
+
+using std::string;
+
```
```suggestion
```

vulder commented on unittests/Feature/FeatureModelBuilder.cpp:
```diff
@@ -34,81 +34,101 @@ TEST(FeatureModelBuilder, addOptionalFeature) {
 }
 
-TEST(FeatureModelBuilder, addExclude) {
+// TODO(s9latimm): Replace string equals with subtree comparison
```
Should this be cleaned up?

vulder commented on unittests/Feature/FeatureModelBuilder.cpp:
```diff
@@ -34,81 +34,101 @@ TEST(FeatureModelBuilder, addOptionalFeature) {
   B.makeFeature<BinaryFeature>("a");
-  B.addExclude("a", "b")->makeFeature<BinaryFeature>("b");
+  B.makeFeature<BinaryFeature>("b");
+  auto C = std::make_unique<ImpliesConstraint>(
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<Feature>("a")),
+      std::make_unique<NotConstraint>(
+          std::make_unique<PrimaryFeatureConstraint>(
+              std::make_unique<Feature>("b"))));
```
Seeing this multiple times, maybe it would make sense to create a ConstraintBuild but we can also do this later in a refactoring. If you want to do it later, could you please create an issue for it :smile:

vulder commented on unittests/Feature/FeatureModelWriter.cpp:
```diff
@@ -37,7 +39,8 @@ TEST(XmlWriter, excludes) {
   std::string ExpectedOutput = FS.get()->getBuffer();
-  EXPECT_EQ(ActualOutput, ExpectedOutput);
+  // TODO(s9latimm): Enable with refactored Constraint
+  //  EXPECT_EQ(ActualOutput, ExpectedOutput);
```
what is broken here?

vulder commented on unittests/Feature/FeatureModelWriter.cpp:
```diff
@@ -54,7 +57,8 @@ TEST(XmlWriter, test) {
   std::string ExpectedOutput = FS.get()->getBuffer();
-  EXPECT_EQ(ActualOutput, ExpectedOutput);
+  // TODO(s9latimm): Enable with refactored Constraint
+  //  EXPECT_EQ(ActualOutput, ExpectedOutput);
```
same I guess?

vulder reviewed (changes requested):
So, I made a full pass over the code. This became quite a pull request :laughing: :+1:

vulder commented on unittests/Feature/FeatureModel.cpp:
```diff
@@ -1,3 +1,4 @@
+#include "llvm/ADT/SetVector.h"
 #include <vara/Feature/FeatureModel.h>
```
```suggestion
#include "vara/Feature/FeatureModel.h"

#include "llvm/ADT/SetVector.h"
```

s9latimm committed:
Extend CK

M	include/vara/Feature/Constraint.h

s9latimm committed:
WIP

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/Constraint.cpp
M	lib/Feature/Feature.cpp
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/Relationship.cpp

s9latimm committed:
Add iterator ranges

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	lib/Feature/Constraint.cpp
M	lib/Feature/Feature.cpp

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -99,6 +105,57 @@ class Feature : public FeatureTreeNode {
+        Excludes.push_back(E);
+      }
+    }
+    return llvm::make_range(Excludes.begin(), Excludes.end());
```
You are returning a iterator to a local stack variable here, `Excludes` scope ends after this functions.
As far as I can tell, this function creates an arbitrary amount of nasal demons.
![image](https://user-images.githubusercontent.com/3064168/99319039-24dbe500-2869-11eb-9ced-3093a25b219e.png)

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -99,6 +105,57 @@ class Feature : public FeatureTreeNode {
+      }
+    }
+    return llvm::make_range(Implications.begin(), Implications.end());
+  }
```
Same with these three cases.

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -154,9 +154,9 @@ class FeatureModelBuilder : private FeatureModel {
-                  const std::vector<std::string> &FeatureNames,
-                  const std::string &ParentName) {
+  emplaceRelationship(Relationship::RelationshipKind RK,
+                      const std::vector<std::string> &FeatureNames,
```
```suggestion
                      std::vector<std::string> FeatureNames,
```

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -154,9 +154,9 @@ class FeatureModelBuilder : private FeatureModel {
+  emplaceRelationship(Relationship::RelationshipKind RK,
+                      const std::vector<std::string> &FeatureNames,
+                      const std::string &ParentName) {
     RelationshipEdges[ParentName].emplace_back(RK, FeatureNames);
```
```suggestion
    RelationshipEdges[ParentName].emplace_back(RK, std::move(FeatureNames));
```

s9latimm committed:
Add mutex

M	include/vara/Feature/Feature.h
M	lib/Feature/Feature.cpp
M	lib/Feature/FeatureModel.cpp

s9latimm committed:
Add algorithm

M	lib/Feature/FeatureModel.cpp

s9latimm committed:
Suggested changes

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureTreeNode.h
M	include/vara/Feature/OrderedFeatureVector.h
M	include/vara/Feature/Relationship.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp

s9latimm committed:
TODO writer

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/Feature.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/FeatureModelWriter.cpp
M	unittests/resources/test.xml

s9latimm committed:
Update shapes

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/FeatureModel.h
M	unittests/resources/test.xml

s9latimm committed:
Update xor detection

M	lib/Feature/FeatureModel.cpp

s9latimm committed:
Add constraint removal

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureTreeNode.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp

s9latimm committed:
Fix writer tests

M	.travis.yml
M	include/vara/Feature/OrderedFeatureVector.h
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/FeatureModelWriter.cpp
M	unittests/resources/CMakeLists.txt
M	unittests/resources/test.xml
M	unittests/resources/test_children.xml
M	unittests/resources/test_excludes.xml
D	unittests/resources/test_out.xml

s9latimm committed:
Fix some bindings, add TODOs

M	bindings/python/tests/TEST_INPUTS/example_feature_model.xml
M	bindings/python/tests/TEST_INPUTS/simple_example_feature_model.xml
M	bindings/python/tests/feature/test_feature.py
M	bindings/python/vara-feature/pybind_Feature.cpp
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureTreeNode.h
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/FeatureModelBuilder.cpp

vulder committed:
Fixes llvm 11 explicit StringRef conversions

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -19,7 +19,7 @@ namespace vara::feature {
 class ConstraintVisitor;
 
-/// \brief Tree like representation of constraints between features
+/// \brief Tree like representation of constraints between features3
```
```suggestion
/// \brief Tree like representation of constraints between features
```
I assume the `3` does not carry meaning :wink:

vulder commented on lib/Feature/FeatureModelWriter.cpp:
```diff
@@ -184,12 +185,25 @@ int FeatureModelXmlWriter::writeBooleanConstraints( // NOLINT
+      V.insert(getChildrenFeatures(C));
+    }
+  }
+  return std::move(V);
```
```suggestion
  return V;
```
If you return a local variable in all return paths C++17 guarantees return-value optimization, which can directly constructs the value in the caller stack frame. Using move here would prevent RVO.

vulder commented on lib/Feature/FeatureModelWriter.cpp:
```diff
@@ -184,12 +185,25 @@ int FeatureModelXmlWriter::writeBooleanConstraints( // NOLINT
+    } else {
+      V.insert(getChildrenFeatures(C));
+    }
+  }
```
I would rewrite this a bit. Use a top level function with the same signature and a second private impl, which takes also a OrderedFeatureVector.
The top level function creates the OFV and passes it to the impl that fill the vector. With this impl you don't have to create a vector for every recursive step and delete it afterwards.
+ the interface for the user is the same :heavy_check_mark:

vulder commented on lib/Feature/FeatureModel.cpp:
```diff
@@ -78,19 +78,44 @@ bool isSimpleMutex(const Feature *A, const Feature *B) {
 
+/// Collect mutual exclusive feature constraints for clean up.
+llvm::SmallSet<const Constraint *, 3>
+cleanUpMutex(const Feature *A, const llvm::SmallSet<Feature *, 3> &Xor) {
```
We could come up with a better name :laughing: My first thought was "why is he using a mutex here"
Maybe `cleanUpMutualExclusiveConstraints` ?

vulder commented on include/vara/Feature/OrderedFeatureVector.h:
```diff
@@ -26,6 +26,12 @@ class OrderedFeatureVector {
   /// Insert feature while preserving ordering.
   void insert(Feature *F);
 
+  void insert(const OrderedFeatureVector &OFV) {
+    for (const auto &F : OFV) {
+      insert(F);
+    }
+  }
+
```
Together with the initializer list insert, we should create an insert that takes two iterators and just forward begin/end to it if we have a container insert like this or initializer list

s9latimm committed:
Some improvements to OVF

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/OrderedFeatureVector.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/OrderedFeatureVector.cpp

s9latimm committed:
Fix tidy

M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/OrderedFeatureVector.cpp

s9latimm committed:
Fix bindings

M	bindings/python/tests/feature/test_feature.py
M	bindings/python/vara-feature/pybind_Feature.cpp

s9latimm committed:
Clean up xml writer

M	include/vara/Feature/FeatureModelWriter.h
M	include/vara/Feature/FeatureTreeNode.h
M	include/vara/Feature/OrderedFeatureVector.h
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/FeatureModelWriter.cpp

s9latimm committed:
LLVM 11 as default

M	CMakeLists.txt

s9latimm committed:
Fix tidy

M	.clang-tidy
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/Constraint.cpp
M	lib/Feature/FeatureModelParser.cpp

s9latimm committed:
Fix tidy

M	.clang-tidy
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/FeatureModel.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
Fix tidy release

M	lib/Feature/FeatureModel.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/NumericFeature.cpp

vulder commented on include/vara/Feature/OrderedFeatureVector.h:
```diff
@@ -13,40 +13,51 @@ namespace vara::feature {
-  OrderedFeatureVector(FeatureIterTy Start, FeatureIterTy End) {
-    insert(llvm::iterator_range(std::move(Start), std::move(End)));
+  OrderedFeatureVector(FeatureIterTy Begin, FeatureIterTy End) {
+    insert(std::move(Begin), std::move(End));
```
```suggestion
    insert(std::forward<FeatureIterTy>(Begin), std::forward<FeatureIterTy>(End));
```
You need to use forward here because you don't not the type `FeatureIterTy`

vulder commented on include/vara/Feature/OrderedFeatureVector.h:
```diff
@@ -13,40 +13,51 @@ namespace vara::feature {
-    insert(FF...);
+  template <class FeatureIterTy>
+  void insert(FeatureIterTy Begin, FeatureIterTy End) {
+    insert(llvm::make_range(Begin, End));
```
```suggestion
    insert(llvm::make_range(std::forward<FeatureIterTy>(Begin), std::forward<FeatureIterTy>(End)));
```

s9latimm committed:
Apply suggestions from code review

Co-authored-by: Florian Sattler <vuld3r@gmail.com>
